pkg/assembler/backends/keyvalue/certifyGood.go

```diff
@@ -191,7 +192,161 @@ func (c *demoClient) ingestCertifyGood(ctx context.Context, subject model.Packag
 // Query CertifyGood
 
 func (c *demoClient) CertifyGoodList(ctx context.Context, certifyGoodSpec model.CertifyGoodSpec, after *string, first *int) (*model.CertifyGoodConnection, error) {
-	return nil, fmt.Errorf("not implemented: CertifyGoodList")
+	c.m.RLock()
+	defer c.m.RUnlock()
+
+	funcName := "CertifyGood"
+
+	if certifyGoodSpec.ID != nil {
+		link, err := byIDkv[*goodLink](ctx, *certifyGoodSpec.ID, c)
+		if err != nil {
+			// Not found
+			return nil, nil
+		}
+
+		foundCertifyGood, err := c.buildCertifyGood(ctx, link, &certifyGoodSpec, true)
+		if err != nil {
+			return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+		}
+
+		return &model.CertifyGoodConnection{
+			TotalCount: 1,
+			PageInfo: &model.PageInfo{
+				HasNextPage: false,
+				StartCursor: ptrfrom.String(foundCertifyGood.ID),
+				EndCursor:   ptrfrom.String(foundCertifyGood.ID),
+			},
+			Edges: []*model.CertifyGoodEdge{
+				{
+					Cursor: foundCertifyGood.ID,
+					Node:   foundCertifyGood,
+				},
+			},
+		}, nil
+	}
+
+	edges := make([]*model.CertifyGoodEdge, 0)
+	hasNextPage := false
+	numNodes := 0
+	totalCount := 0
+
+	// Cant really search for an exact Pkg, as these can be linked to either
+	// names or versions, and version could be empty.
+	var search []string
+	foundOne := false
+	if certifyGoodSpec.Subject != nil && certifyGoodSpec.Subject.Artifact != nil {
+		exactArtifact, err := c.artifactExact(ctx, certifyGoodSpec.Subject.Artifact)
+		if err != nil {
+			return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+		}
+		if exactArtifact != nil {
+			search = append(search, exactArtifact.GoodLinks...)
+			foundOne = true
+		}
+	}
+	if !foundOne && certifyGoodSpec.Subject != nil && certifyGoodSpec.Subject.Source != nil {
+		exactSource, err := c.exactSource(ctx, certifyGoodSpec.Subject.Source)
+		if err != nil {
+			return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+		}
+		if exactSource != nil {
+			search = append(search, exactSource.GoodLinks...)
+			foundOne = true
+		}
+	}
+
+	if foundOne {
+		for _, id := range search {
+			link, err := byIDkv[*goodLink](ctx, id, c)
+			if err != nil {
+				return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+			}
+			cg, err := c.CGIfMatch(ctx, &certifyGoodSpec, link)
```

Comment:
need to do these for all these for both the pagination and regular query.

Reply:
Added the nil check for all the pagination and regular queries.